include/range/v3/numeric/partial_sum.hpp

```diff
@@ -30,16 +30,14 @@ namespace ranges
     inline namespace v3
     {
         template<typename I, typename O, typename BOp = plus, typename P = ident,
-            typename V = value_type_t<I>,
-            typename X = concepts::Invocable::result_t<P&, V>,
-            typename Y = concepts::Invocable::result_t<BOp&, X, X>>
+            typename X = projected<projected<I, coerce<value_type_t<I>>>, P>,
+            typename Y = reference_t<X>>
         using PartialSummable = meta::strict_and<
+            SemiRegular<detail::decay_t<Y>>,
```

Comment:
It seems like trying to constrain the behavior to avoid such 'evil' types is impossible. How would you check for lack of side effects? Perhaps it makes no sense for range library to try and prevent all the potential abuses?